File: app/src/main/java/com/shetj/diyalbume/view/AlbumImageView.java
package com.shetj.diyalbume.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathEffect;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.util.AttributeSet;
import android.view.MotionEvent;

import com.shetj.diyalbume.R;

/**
 * <b>@packageName：</b> com.shetj.diyalbume.view<br>
 * <b>@author：</b> shetj<br>
 * <b>@createTime：</b> 2017/12/19<br>
 * <b>@company：</b><br>
 * <b>@email：</b> 375105540@qq.com<br>
 * <b>@describe</b><br>
 */

public class AlbumImageView extends AppCompatImageView{

	public static final int DEFAULT_FRAME_COLOR = Color.RED;


	private Drawable controlDrawable;//移动图片
	private Drawable deleteDrawable;//删除图片


	private Paint mPaint;

	private Path mPath = new Path();//外框

	private int mFrameColor = DEFAULT_FRAME_COLOR ;
	private boolean isDrag = false;


	public AlbumImageView(Context context) {
		super(context);
		init();
	}

	public AlbumImageView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		init();
	}

	public AlbumImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		obtainStyledAttributes(attrs);
	}

	private void obtainStyledAttributes(AttributeSet attrs) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.AlbumImageView);

	}

	private void init() {
		mPaint = new Paint();
		mPaint.setColor(mFrameColor);//设置画笔颜色
		mPaint.setAntiAlias(true);//设置抗锯齿
		mPaint.setStyle(Paint.Style.STROKE);
		PathEffect pathEffect = new DashPathEffect(new float[] { 15, 10 }, 5);
		mPaint.setPathEffect(pathEffect);



	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		mPath.reset();
		mPath.moveTo(0,0);
		mPath.lineTo(getWidth(),0);
		mPath.lineTo(getWidth(),getHeight());
		mPath.lineTo(0,getHeight());
		mPath.lineTo(0,0);
		canvas.drawPath(mPath,mPaint);
	}

	private float lastX;
	private float lastY;

	@Override
	public boolean onTouchEvent(MotionEvent event) {
		float x = event.getRawX();
		float y = event.getRawY();
		switch (event.getAction()){
			case MotionEvent.ACTION_DOWN:
				isDrag = false;
				lastX = x;
				lastY = y;
				break;

			case MotionEvent.ACTION_MOVE:
				isDrag = true;
				float dx=x-lastX;
				float dy=y-lastY;

				float x1 = getX() + dx;
				float y1 = getY() + dy;

				setX(x1);
				setY(y1);

				lastX=x;
				lastY=y;
				break;
			case MotionEvent.ACTION_UP:
				break;
		}
		return  !isNotDrag() || super.onTouchEvent(event);
	}

	private boolean isNotDrag() {
		return  !isDrag;
	}

	public String getJson() {
		return "{"+getX()+","+getY()+"}";
	}
}

File: app/src/main/java/com/shetj/diyalbume/view/AlbumImageView.kt
package com.shetj.diyalbume.view

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.annotation.RequiresApi
import androidx.appcompat.widget.AppCompatImageView
import com.shetj.diyalbume.R

/**
 * **@packageName：** com.shetj.diyalbume.view<br></br>
 * **@author：** shetj<br></br>
 * **@createTime：** 2017/12/19<br></br>
 * **@company：**<br></br>
 * **@email：** 375105540@qq.com<br></br>
 * **@describe**<br></br>
 */
@RequiresApi(Build.VERSION_CODES.HONEYCOMB)
class AlbumImageView : AppCompatImageView {


    private val controlDrawable: Drawable? = null//移动图片
    private val deleteDrawable: Drawable? = null//删除图片


    private var mPaint: Paint? = null

    private val mPath = Path()//外框

    private val mFrameColor = DEFAULT_FRAME_COLOR
    private var isDrag = false

    private var lastX: Float = 0.toFloat()
    private var lastY: Float = 0.toFloat()

    private val isNotDrag: Boolean
        get() = !isDrag

    val json: String
        get() = "{$x,$y}"


    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        obtainStyledAttributes(attrs)
    }

    private fun obtainStyledAttributes(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.AlbumImageView)

    }

    private fun init() {
        mPaint = Paint()
        mPaint!!.color = mFrameColor//设置画笔颜色
        mPaint!!.isAntiAlias = true//设置抗锯齿
        mPaint!!.style = Paint.Style.STROKE
        val pathEffect = DashPathEffect(floatArrayOf(15f, 10f), 5f)
        mPaint!!.pathEffect = pathEffect
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        mPath.reset()
        mPath.moveTo(0f, 0f)
        mPath.lineTo(width.toFloat(), 0f)
        mPath.lineTo(width.toFloat(), height.toFloat())
        mPath.lineTo(0f, height.toFloat())
        mPath.lineTo(0f, 0f)
        canvas.drawPath(mPath, mPaint!!)

    }


    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.rawX
        val y = event.rawY
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                isDrag = false
                lastX = x
                lastY = y
            }

            MotionEvent.ACTION_MOVE -> {
                isDrag = true
                val dx = x - lastX
                val dy = y - lastY
                val x1 = getX() + dx
                val y1 = getY() + dy
                setX(x1)
                setY(y1)
                lastX = x
                lastY = y
            }
            MotionEvent.ACTION_UP -> {
            }
            else -> {
            }
        }
        return !isNotDrag || super.onTouchEvent(event)
    }

    companion object {

        val DEFAULT_FRAME_COLOR = Color.RED
    }
}
